### PR TITLE
SYS-2757 Track collator payouts

### PR DESCRIPTION
**Pull request by nahuseyoum**
This PR adds the required data structure and logic to track collator reward payouts which will be done via the T1 growth process.

### CODE REVIEW

File: pallets/parachain-staking/src/types.rs [hunk 1]
```diff
@@ -1566,3 +1567,39 @@ pub enum BondAdjust<Balance> {
     Increase(Balance),
     Decrease,
 }
+
+// Data structure for tracking collator rewards
+#[derive(Clone, PartialEq, Eq, Encode, Decode, RuntimeDebug, Default, TypeInfo)]
+pub struct CollatorPayoutInfo<Balance> {
+    pub total_stake_accumulator: CollatorPayoutPeriodIndex,
```
Comment (alexpArtos) [inline pallets/parachain-staking/src/types.rs:1574]: I'd suggest some different naming here.
```suggestion
    pub number_of_accumulations: CollatorPayoutPeriodIndex,
```

File: pallets/parachain-staking/src/types.rs [hunk 2]
```diff
@@ -1566,3 +1567,39 @@ pub enum BondAdjust<Balance> {
...
+    pub total_amount_staked: Balance,
```
Comment (alexpArtos) [inline pallets/parachain-staking/src/types.rs:1575]: And here:
```suggestion
    pub total_stake_accumulated: Balance,
```

File: pallets/parachain-staking/src/lib.rs [hunk 3]
```diff
@@ -1496,6 +1518,49 @@ pub mod pallet {
             T::Currency::free_balance(&Self::compute_reward_pot_account_id())
                 .saturating_sub(T::Currency::minimum_balance())
         }
+
+        pub fn update_collator_payout(
+            new_era: EraIndex,
+            total_staked: BalanceOf<T>,
```
Comment (alexpArtos) [inline pallets/parachain-staking/src/lib.rs:1524]:
```suggestion
            total_staked_at_era: BalanceOf<T>,
```

File: pallets/parachain-staking/src/lib.rs [hunk 4]
```diff
@@ -1496,6 +1518,49 @@ pub mod pallet {
...
+            payout: DelayedPayout<BalanceOf<T>>,
+        ) {
+            let collator_payout_period = Self::collator_payout_period_info();
+            let new_total_staking_reward = payout.total_staking_reward;
```
Comment (alexpArtos) [inline pallets/parachain-staking/src/lib.rs:1528]:
```suggestion
            let staking_reward_paid_in_era = payout.total_staking_reward;
```

File: pallets/parachain-staking/src/lib.rs [hunk 5]
```diff
@@ -1496,6 +1518,49 @@ pub mod pallet {
...
+
+            if Self::is_new_payout_period(&new_era, &collator_payout_period) {
+                <CollatorPayoutPeriod<T>>::mutate(|info| {
```
Comment (alexpArtos) [inline pallets/parachain-staking/src/lib.rs:1531]: Please, extract this into a function. Possible name: `recordNewPayoutPeriod`.

File: pallets/parachain-staking/src/lib.rs [hunk 6]
```diff
@@ -1496,6 +1518,49 @@ pub mod pallet {
...
+                    Self::collator_payout_period_info().index,
+                    new_payout_info,
+                );
+            } else {
+                <CollatorPayout<T>>::mutate(collator_payout_period.index, |info| {
+                    info.total_stake_accumulator = info.total_stake_accumulator.saturating_add(1);
+                    info.total_amount_staked =
+                        info.total_amount_staked.saturating_add(total_staked);
+                    info.total_staker_reward =
+                        info.total_staker_reward.saturating_add(new_total_staking_reward);
```
Comment (alexpArtos) [inline pallets/parachain-staking/src/lib.rs:1551]: Please extract this into a function. Possible name: `accumulateEraInPayoutPeriod`.

**Review by alexpArtos — APPROVED** [review]

**Review by fluorostani — APPROVED** [review]

**Review by thadouk — APPROVED** [review]

**Pull request merged**